test/performance/tiledb_submit_a_b.cc

```diff
@@ -0,0 +1,305 @@
+    tiledb_ctx_t* ctx,
+    tiledb_array_t* array,
+    tiledb_query_t* query,
+    const json& jq) {
+  if (jq.find("subarray") != jq.end()) {
+    tiledb_subarray_t* subarray;
+    RETURN_IF_ERR(tiledb_subarray_alloc(ctx, array, &subarray));
+
+    tiledb_array_schema_t* schema;
+    RETURN_IF_ERR(tiledb_array_get_schema(ctx, array, &schema));
+
+    tiledb_domain_t* domain;
+    RETURN_IF_ERR(tiledb_array_schema_get_domain(ctx, schema, &domain));
+
+    int d = 0;
+    const auto& jsub = jq["subarray"];
+    for (const auto& jdim : jsub) {
+      if (jdim.find("%") != jdim.end()) {
+        const int percent = jdim["%"].get<int>();
+
+        // FIXME: make generic
+        int is_empty;
+        int64_t domain[2];
+        RETURN_IF_ERR(tiledb_array_get_non_empty_domain_from_index(
+            ctx, array, d, &domain[0], &is_empty));
+        if (is_empty) {
+          continue;
+        }
+
+        const auto span = (domain[1] - domain[0] + 1);
+        domain[0] += ((span * (100 - percent)) / 100) / 2;
+        domain[1] -= ((span * (100 - percent)) / 100) / 2;
+
+        RETURN_IF_ERR(tiledb_subarray_add_range(
+            ctx, subarray, d, &domain[0], &domain[1], 0));
+      }
+    }
+
+    tiledb_query_set_subarray_t(ctx, query, subarray);
+  }
+
+  return TILEDB_OK;
+}
+
+template <FragmentType Fragment>
+static void run(
+    TimeKeeper& time_keeper,
+    const char* array_uri,
+    const json& query_config,
+    tiledb_config_t* a_config,
+    tiledb_config_t* b_config) {
+  tiledb::test::SparseGlobalOrderReaderMemoryBudget memory;
+  memory.total_budget_ = std::to_string(1024 * 1024 * 1024);
+  memory.ratio_tile_ranges_ = "0.01";
+
+  const uint64_t num_user_cells = 1024 * 1024 * 128;
+
+  Fragment a;
+  Fragment b;
+
+  auto reset = [num_user_cells](Fragment& buf) {
+    std::apply(
+        [&](auto&... outfield) { (outfield.resize(num_user_cells), ...); },
+        std::tuple_cat(buf.dimensions(), buf.attributes()));
+  };
+  reset(a);
+  reset(b);
+
+  tiledb_config_t* config;
+  {
+    tiledb_error_t* error = nullptr;
+    ASSERTER(tiledb_config_alloc(&config, &error) == TILEDB_OK);
+  }
+  memory.apply(config);
+
+  tiledb_ctx_t* ctx;
+  ASSERTER(tiledb_ctx_alloc(config, &ctx) == TILEDB_OK);
+
+  tiledb_config_free(&config);
+
+  // Open array for reading.
+  CApiArray array(ctx, array_uri, TILEDB_READ);
+
+  auto dimension_name = [&](unsigned d) -> std::string {
+    return std::string(
+        array->array_schema_latest().domain().dimension_ptr(d)->name());
+  };
+  auto attribute_name = [&](unsigned a) -> std::string {
+    return std::string(array->array_schema_latest().attribute(a)->name());
+  };
+
+  // Create query which does NOT do merge
+  tiledb_query_t* a_query;
+  TRY(ctx, tiledb_query_alloc(ctx, array, TILEDB_READ, &a_query));
+  TRY(ctx, tiledb_query_set_layout(ctx, a_query, TILEDB_GLOBAL_ORDER));
+  TRY(ctx, tiledb_query_set_config(ctx, a_query, a_config));
+  json2query(ctx, array, a_query, query_config);
```

Comment:
NIT: that function name was a little confusing for me, something like `set_subarray_from_json` would be clearer to the reader

Reply:
Changed to `construct_query` in 88b11c